### PR TITLE
Claim hooks pt 1 - useClaimCallback

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Updated `ClaimType` from a TS type to an enum, to be inline with contract
Added modified `useClaimCallback`

Example of a successful tx claiming `1080` https://rinkeby.etherscan.io/tx/0x90ad7c1ffc412c75286f91ae821c23397bbe0e2e75fdcf6e5961101c7b1f220e

Screenshot from 2021-12-24 07-16-03

  # To Test

0. Edit [this file](https://github.com/gnosis/cowswap/pull/2041/files#diff-3ae56d49d1cbb7b1aea829a94599d8ae0f20f9c5ec804ad13ad50b9a92ac31c8R72) with a different claim since I already claimed `1080` (no way to do that in the UI yet)
1. Connect wallet to rinkeby
2. Click on `claim vCOW`
3. Click on `claim` in the modal
* Your transaction should be sent

  # Background

*Optional: Give background information for changes you've made, that might be difficult to explain via comments*

### CODE REVIEW

File: src/components/claim/ClaimModal.tsx [hunk 1]
```diff
@@ -51,23 +51,25 @@ export default function ClaimModal() {
...
+  const { claimCallback } = useClaimCallback(account) // TODO: remove me, hard coded only for testing
   const unclaimedAmount: CurrencyAmount<Token> | undefined = useUserUnclaimedAmount(account)
   const { claimSubmitted, claimTxn } = useUserHasSubmittedClaim(account ?? undefined)
   const claimConfirmed = Boolean(claimTxn?.receipt)
 
   function onClaim() {
+    console.log(`Trying to claim!!!`, unclaimedAmount?.toString(), claimConfirmed)
     setAttempting(true)
-    claimCallback()
+    claimCallback([{ index: 1080 }])
```
Comment (alfetopito) [inline src/components/claim/ClaimModal.tsx:72]: Trying to claim the claim `1080` for account `0x0Ae21f9700eD15E1a83baa3183a67d54b1DA2086`

File: src/custom/state/claim/hooks/index.ts [hunk 2]
```diff
@@ -94,3 +113,178 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+          !connectedAccount,
+          !chainId,
+          !vCowContract,
+          !vCowToken
+        )
+        throw new Error("Not initialized, can't claim")
+      }
+      console.log(`claimHooks::claimCallback::got through`, claims, claimInput, account, connectedAccount, chainId)
+
+      const { args, totalClaimedAmount, value } = _buildArgs({ claimInput, claims, account, connectedAccount })
```
Comment (alfetopito) [inline src/custom/state/claim/hooks/index.ts:242]: Args at this point looks like this:

```json
[
  [
    1080
  ],
  [
    0
  ],
  [
    "0x0Ae21f9700eD15E1a83baa3183a67d54b1DA2086"
  ],
  [
    "9289000000000000000000"
  ],
  [
    "9289000000000000000000"
  ],
  [
    [
      "0x03cc3df79d865ad3bfbdd54ec8ee6001f294655fe0018e660c0879c046d82133",
      "0xd044fdd7448e035e36a13a1b3edb2d7158624c6578a31517e8deefc2034c216b",
      "0x4531947895cb7cfc7d878e32288a73c04eaf63af8345d47283ba2cb8bac7e8c5",
      "0xf1376d3d11b33948440cf71bcfa63e69a9ca436ab3cef016b9da36f2a226d3e2",
      "0xa5ff51dde0385f442354e8d2582e81fd7d81f0a9b558ceea8e84136e6b88501c",
      "0xdc13f692869b72e45579e0ef7f5afe670ffeb5c2931f5b72cc50a19121c57f68",
      "0xbf454c5b68f2a3637a1221ce1e5834ce5725f9fcef7f932f81af24127e312d58",
      "0x0c85968537348d6281fb14be6b4528e77fce7adb1ac9a6fae52c83182f64d848",
      "0xdc838baad0c5b2844072e96cd5376ca337f80b42746b46a46121afcc0e39e0ee",
      "0x52aaac536243500f3d5aae091887032322773b40f7fa906d0043a5d74013a8c4",
      "0x57158a64c297f7a0ae2530774437f91667aab006e084f3214986dce3ca898474",
      "0xc1d3779a02affbc6b4885b248a636f9a3fd124d356e639347bb04129436133bd",
      "0xe0b4fbf77c444bada1498ecbee1431eb297d4fa9a6d04f1dcb19571c334b5dcb",
      "0x768e2f52bd7f4c8f1251620800bed30d14d13453dd2f0ced98eff0212d5e5288",
      "0x3fb9f3c87451effbb58340761f30e2d2057a0a2fb27eee5ec6af3362be0389f9"
    ]
  ],
  [
    "0"
  ]
]
```

File: src/custom/state/claim/hooks/index.ts [hunk 3]
```diff
@@ -94,3 +113,178 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+      //   .then((response: TransactionResponse) => {
+      //     addTransaction({
+      //       hash: response.hash,
+      //       summary: `Claimed ${formatSmart(vCowAmount)} vCOW`,
+      //       claim: { recipient: account },
+      //     })
+      //     return response.hash
+      //   })
+
+      return vCowContract.estimateGas['claimMany'](...args, { value }).then((estimatedGas) => {
```
Comment (alfetopito) [inline src/custom/state/claim/hooks/index.ts:268]: This call fails with the error:

Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"code":-32603,"message":"execution reverted","data":{"originalError":{"code":3,"data":"0x09bde339","message":"execution reverted"}}}, method="estimateGas", transaction={"from":"0x5B0ABE214aB7875562ADeE331dEfF0Fe1912FE42","to":"0x8F07778b703F9FEB93521FC5230e859Dbd032b6b","data":"0xf3507dab00000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000001a000000000000000000000000000000000000000000000000000000000000001e000000000000000000000000000000000000000000000000000000000000002200000000000000000000000000000000000000000000000000000000000000460000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000004380000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000ae21f9700ed15e1a83baa3183a67d54b1da208600000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000001f78ec55cc4cb84000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000001f78ec55cc4cb84000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000f03cc3df79d865ad3bfbdd54ec8ee6001f294655fe0018e660c0879c046d82133d044fdd7448e035e36a13a1b3edb2d7158624c6578a31517e8deefc2034c216b4531947895cb7cfc7d878e32288a73c04eaf63af8345d47283ba2cb8bac7e8c5f1376d3d11b33948440cf71bcfa63e69a9ca436ab3cef016b9da36f2a226d3e2a5ff51dde0385f442354e8d2582e81fd7d81f0a9b558ceea8e84136e6b88501cdc13f692869b72e45579e0ef7f5afe670ffeb5c2931f5b72cc50a19121c57f68bf454c5b68f2a3637a1221ce1e5834ce5725f9fcef7f932f81af24127e312d580c85968537348d6281fb14be6b4528e77fce7adb1ac9a6fae52c83182f64d848dc838baad0c5b2844072e96cd5376ca337f80b42746b46a46121afcc0e39e0ee52aaac536243500f3d5aae091887032322773b40f7fa906d0043a5d74013a8c457158a64c297f7a0ae2530774437f91667aab006e084f3214986dce3ca898474c1d3779a02affbc6b4885b248a636f9a3fd124d356e639347bb04129436133bde0b4fbf77c444bada1498ecbee1431eb297d4fa9a6d04f1dcb19571c334b5dcb768e2f52bd7f4c8f1251620800bed30d14d13453dd2f0ced98eff0212d5e52883fb9f3c87451effbb58340761f30e2d2057a0a2fb27eee5ec6af3362be0389f900000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000","accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.4.5)

File: src/components/claim/ClaimModal.tsx [hunk 4]
```diff
@@ -51,23 +51,25 @@ export default function ClaimModal() {
   const isOpen = useModalOpen(ApplicationModal.SELF_CLAIM)
   const toggleClaimModal = useToggleSelfClaimModal()
 
-  const { account, chainId } = useActiveWeb3React()
+  const { chainId } = useActiveWeb3React()
 
+  const account = '0x0Ae21f9700eD15E1a83baa3183a67d54b1DA2086'
```
Comment (alfetopito) [inline src/components/claim/ClaimModal.tsx:56]: Temporary hard coded changes

File: src/custom/constants/index.ts [hunk 5]
```diff
@@ -57,7 +57,7 @@ export const V_COW_CONTRACT_ADDRESS: Partial<Record<number, string>> = {
   // [ChainId.MAINNET]: GPv2Settlement[ChainId.MAINNET].address,
   // [ChainId.RINKEBY]: GPv2Settlement[ChainId.RINKEBY].address,
   // [ChainId.XDAI]: GPv2Settlement[ChainId.XDAI].address,
-  [ChainId.RINKEBY]: '0x8f07778b703f9feb93521fc5230e859dbd032b6b',
+  [ChainId.RINKEBY]: '0xac17Ae2BcAF9Ed181C4Ab10B179548d6aa30C92E',
```
Comment (alfetopito) [inline src/custom/constants/index.ts:60]: This is why it didn't work before, I was using the wrong contract address...

File: src/components/claim/ClaimModal.tsx (hunk 1 above)
Comment (W3stside) [reply src/components/claim/ClaimModal.tsx:72]: use registerOnWindow and make a global index changeable?

File: src/custom/constants/index.ts (hunk 5 above)
Comment (W3stside) [reply src/custom/constants/index.ts:60]: https://media4.giphy.com/media/xT9DPJVjlYHwWsZRxm/giphy.gif?cid=790b76118dea2e68410684f9b11b5c4ac93fa1130bf1af15&rid=giphy.gif&ct=g

File: src/custom/state/claim/hooks/index.ts [hunk 6]
```diff
@@ -1,25 +1,33 @@
...
 export * from './hooksMod'
 
-export type ClaimType =
-  | 'Airdrop' // free, no vesting, can be available on both mainnet and gchain
-  | 'Team' // free, with vesting, only on mainnet
-  | 'Advisor' // free, with vesting, only on mainnet
-  | 'GnoOption' // paid, with vesting, must use GNO, can be available on both mainnet and gchain
-  | 'UserOption' // paid, with vesting, must use Native currency, can be available on both mainnet and gchain
-  | 'Investor' // paid, with vesting, must use USDC, only on mainnet
+export enum ClaimType {
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:20]: haaaa yes enums lol. export const enum? wasn't that smaller or sth

File: src/custom/state/claim/hooks/index.ts [hunk 7]
```diff
@@ -94,3 +114,160 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+
+type GetClaimManyArgsParams = {
+  claimInput: ClaimInput[]
+  claims: UserClaims
+  account: string
+  connectedAccount: string
+}
+
+function _getClaimManyArgs({ claimInput, claims, account, connectedAccount }: GetClaimManyArgsParams): {
+  args: [number[], ClaimType[], string[], string[], string[], string[][], string[]] | undefined
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:191]: can u make this a type outside the fn please, also `args`

File: src/components/claim/ClaimModal.tsx (hunk 1 above)
Comment (alfetopito) [reply src/components/claim/ClaimModal.tsx:72]: Don't think it's worth the effort, we'll soon wire this up with the UI

File: src/custom/state/claim/hooks/index.ts [hunk 8]
```diff
@@ -94,3 +114,160 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+  totalClaimedAmount: JSBI
+  value: string | undefined
+} {
+  const indices: number[] = []
+  const claimTypes: ClaimType[] = []
+  const claimants: string[] = []
+  const claimableAmounts: string[] = []
+  const claimedAmounts: string[] = []
+  const merkleProofs: string[][] = []
+  const sendEth: string[] = []
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:201]: can you comment on what theese are ? e.g what is sendEth?

File: src/custom/state/claim/hooks/index.ts [hunk 9]
```diff
@@ -94,3 +114,160 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+
+      // sum of claimedAmounts for the toast notification
+      totalClaimedAmount = JSBI.add(totalClaimedAmount, JSBI.BigInt(claimedAmount))
+      // sum of Native currency to be used on call options
+      totalValue = JSBI.add(totalValue, JSBI.BigInt(value))
+    }
+  })
+
+  return {
+    args:
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:242]: cache this on line 240

File: src/custom/state/claim/hooks/index.ts [hunk 10]
```diff
@@ -94,3 +114,160 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+      indices.length > 0
+        ? [indices, claimTypes, claimants, claimableAmounts, claimedAmounts, merkleProofs, sendEth]
+        : undefined,
+    totalClaimedAmount,
+    value: totalValue.toString() === '0' ? undefined : totalValue.toString(),
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:247]: also cache this

File: src/custom/state/claim/hooks/index.ts (hunk 6 above)
Comment (alfetopito) [reply src/custom/state/claim/hooks/index.ts:20]: Just to be inline with the contract, which uses enums, which are translated as 0-indexed integers and I would have to have a map if using types :shrug: 

Apparently there's no definition or something https://thisthat.dev/const-enum-vs-enum/
Will update

File: src/custom/state/claim/hooks/index.ts [hunk 11]
```diff
@@ -94,3 +114,160 @@ export function useUserUnclaimedAmount(account: string | null | undefined): Curr
...
+}
+
+type GetClaimedAmountParams = Pick<GetClaimManyArgsParams, 'account' | 'connectedAccount'> & {
+  claim: UserClaimData
+  input: ClaimInput
+}
+
+function _getClaimedAmount({ claim, input, account, connectedAccount }: GetClaimedAmountParams): string {
+  if (
+    // claiming a paid claim on behalf of someone else?
```
Comment (W3stside) [inline src/custom/state/claim/hooks/index.ts:258]: i would cache these as well with names specific to their state, like `isClaimingForOther = ....`

**Review by W3stside — APPROVED** [review]
nice

File: src/custom/state/claim/hooks/index.ts (hunk 8 above)
Comment (alfetopito) [reply src/custom/state/claim/hooks/index.ts:201]: Added a comment pointing to where they came from: https://github.com/gnosis/gp-v2-token/blob/main/src/contracts/mixins/MerkleDistributor.sol#L123

**Pull request merged**